blob/ribs_store.go

```diff
@@ -0,0 +1,253 @@
+	"github.com/ipfs/go-cid"
+	"github.com/lotus-web3/ribs"
+	"github.com/lotus-web3/ribs/rbdeal"
+	"github.com/multiformats/go-multihash"
+)
+
+// TODO parameterize this.
+const ribsStoreChunkSize = 4 << 20 // 4 MiB
+
+var (
+	_ Store             = (*RibsStore)(nil)
+	_ io.ReadSeekCloser = (*ribsStoredBlobReader)(nil)
+)
+
+type (
+	// RibsStore is an experimental Store implementation that uses RIBS.
+	// See: https://github.com/filcat/ribs
+	RibsStore struct {
+		ribs    ribs.RIBS
+		maxSize int
+		//index   map[uuid.UUID]*ribsStoredBlob // TODO persist this on disk
+		indexDir string
+	}
+	ribsStoredBlob struct {
+		*Descriptor
+		Chunks []cid.Cid `json:"chunks"`
+	}
+	ribsStoredBlobReader struct {
+		sess   ribs.Session
+		blob   *ribsStoredBlob
+		offset int64
+
+		currentChunkIndex       int
+		currentChunkReader      *bytes.Reader
+		currentChunkPendingSeek int64
+	}
+)
+
+// NewRibsStore instantiates a new experimental RIBS store.
+func NewRibsStore(dir string) (*RibsStore, error) {
+	rbdealDir := path.Join(path.Clean(dir), "rbdeal")
+	if err := os.Mkdir(rbdealDir, 0750); err != nil {
+		return nil, fmt.Errorf("failed to create internal directories: %w", err)
+	}
+	indexDir := path.Join(path.Clean(dir), "index")
+	if err := os.Mkdir(indexDir, 0750); err != nil {
+		return nil, fmt.Errorf("failed to create internal directories: %w", err)
+	}
+
+	// TODO Path to wallet is hardcoded in RIBS. Parameterise it and allow user to configure
+	//      See: https://github.com/FILCAT/ribs/blob/7c8766206ec1e5ec30c613cde2b3a49d0ccf25d0/rbdeal/ribs.go#L156
+
+	rbs, err := rbdeal.Open(rbdealDir)
+	if err != nil {
+		return nil, err
+	}
+	return &RibsStore{
+		ribs:    rbs,
+		maxSize: 32 << 30, // 32 GiB
+		//index:   map[uuid.UUID]*ribsStoredBlob{},
+		indexDir: indexDir,
+	}, nil
+
+}
+
+func (r *RibsStore) Start(_ context.Context) error {
+	// TODO: change RIBS to take context.
+	return r.ribs.Start()
+}
+func (r *RibsStore) Put(ctx context.Context, in io.ReadCloser) (*Descriptor, error) {
+
+	// Generate ID early to fail early if generation fails.
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, err
+	}
+	modtime := time.Now().UTC()
+
+	// TODO incorporate https://github.com/filecoin-project/data-prep-tools/tree/main/docs/best-practices
+	//      also see: https://github.com/anjor/anelace
+	// TODO we could do things here to make commp etc. more efficient.
+	//      for now this implementation remains highly experimental and optimised for velocity.
+	batch := r.ribs.Session(ctx).Batch(ctx)
+
+	splitter := chunk.NewSizeSplitter(in, ribsStoreChunkSize)
+
+	// TODO: Store the byte ranges for satisfying io.ReadSeaker in case chunk size is not constant across blocks?
+	var chunkCids []cid.Cid
+	var size int
+SplitLoop:
+	for {
+		b, err := splitter.NextBytes()
+		switch err {
+		case io.EOF:
+			break SplitLoop
+		case nil:
+			size += len(b)
+			if size > r.maxSize {
+				return nil, ErrBlobTooLarge
+			}
+			blk := blocks.NewBlock(b)
```

Comment:
This will make a CidV0 -- which is pretty deprecated. You probably want cid.Prefix.Sum() on the bytes followed by blocks.NewBlockWithCid